src/main/scala/com/actian/spark_vector/datastream/DataStreamClient.scala

```diff
@@ -47,17 +51,36 @@ case class DataStreamClient(vectorProps: VectorConnectionProperties, table: Stri
   private def prepareUnloadSql(table: String) = s"prepare for x100 stream from $table"
   private def startUnloadSql(selectQuery: String) = s"insert into external table $selectQuery"
 
-  private def executeSql(sql: String): Future[Int] = Future { jdbc.executeStatement(sql) }
+  /** Execute a sql (statement) within a future task */
+  private def executeSql(sql: String): Future[Int] = Future { jdbc.executeStatement(sql) } match {
+    case f =>
+      f onFailure {
+        case t =>
+          logError(t.stackTrace)
+          // FIXME: use rollback() instead when Vector will actually close the DataStreams after unrolling
+          close()
+      }
+      f
+  }
 
   /** The `JDBC` connection used by this client to communicate with `Vector` */
   def getJdbc(): VectorJDBC = jdbc
 
+  /**
+   * Abort sending data to Vector rolling back the open transaction
+   * @note the `JDBC` connection is not closed
+   */
+  def rollback(): Unit = synchronized(if (!jdbc.isClosed) {
```

Comment:
hmm...it works as above, w/o parenthesis, if the "if" condition is on a newline but not on the same one with "synchronized"